File: examples/java/IamAttachPolicy.java
// Asigna (adjunta) una política IAM a un usuario.
// Uso: java IamAttachPolicy <nombre-usuario> <arn-politica>
import java.net.URI;
import java.util.stream.Collectors;
import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.iam.IamClient;

public class IamAttachPolicy {
  public static void main(String[] args) {
    if (args.length < 2) {
      throw new IllegalArgumentException("Uso: java IamAttachPolicy <nombre-usuario> <arn-politica>");
    }
    String userName = args[0];
    String policyArn = args[1];

    IamClient iam = IamClient.builder()
        .endpointOverride(URI.create("http://localhost:4566"))
        .region(Region.US_EAST_1)
        .credentialsProvider(
            StaticCredentialsProvider.create(AwsBasicCredentials.create("test", "test")))
        .build();

    try {
      iam.attachUserPolicy(req -> req.userName(userName).policyArn(policyArn));
      System.out.println("Política " + policyArn + " asignada a " + userName + ".");

      var attached = iam.listAttachedUserPolicies(req -> req.userName(userName));
      String names = attached.attachedPolicies().stream()
          .map(p -> p.policyName())
          .collect(Collectors.joining(", "));
      System.out.println("Políticas actualmente asignadas: " + names);
    } finally {
      iam.close();
    }
  }
}

File: examples/python/iam_attach_policy.py
"""Asigna (adjunta) una política IAM a un usuario.

Uso: python3 iam_attach_policy.py <nombre-usuario> <arn-politica>
"""
import sys

import boto3

iam = boto3.client(
    "iam",
    endpoint_url="http://localhost:4566",
    aws_access_key_id="test",
    aws_secret_access_key="test",
    region_name="us-east-1",
)


def main():
    if len(sys.argv) < 3:
        raise SystemExit("Uso: python3 iam_attach_policy.py <nombre-usuario> <arn-politica>")
    user_name, policy_arn = sys.argv[1], sys.argv[2]

    iam.attach_user_policy(UserName=user_name, PolicyArn=policy_arn)
    print(f"Política {policy_arn} asignada a {user_name}.")

    attached = iam.list_attached_user_policies(UserName=user_name)
    names = [p["PolicyName"] for p in attached["AttachedPolicies"]]
    print("Políticas actualmente asignadas:", names)


if __name__ == "__main__":
    main()
